### ISSUE TITLE
SNMP request fails to resolve hostname if DNS entry is longer than 64 characters

### ISSUE THREAD

**Issue by tpelado** [net-snmp/net-snmp#476]
Good day all,

I'm running into an issue upon updating my net-snmp package. As far as I know, this doesn't happen in 5.8.

```
snmpstatus -V                                                                            
NET-SNMP version: 5.9.1
``` 

The issue is as follows : 

- When snmp tries to resolve a domain that has more than 64 characters, it fails and I get this error : 

``` 
snmpstatus -v2c -c public aaaaaaaaaaaaaaaaaaaaaaaaa.aaaaaaaaaaaaaaaaaaaaaaaaa.taledo.games
zsh: correct 'public' to 'Public' [nyae]? n
snmpstatus: Unknown host (aaaaaaaaaaaaaaaaaaaaaaaaa.aaaaaaaaaaaaaaaaaaaaaaaaa.taledo.games) (No such file or directory)
``` 
(don't mind zsh being silly)

Pinging this domain works : 

``` 
ping aaaaaaaaaaaaaaaaaaaaaaaaa.aaaaaaaaaaaaaaaaaaaaaaaaa.taledo.games                                                                 
PING aaaaaaaaaaaaaaaaaaaaaaaaa.aaaaaaaaaaaaaaaaaaaaaaaaa.taledo.games(2001:41d0:801:2000::539f) 56 data bytes
64 bytes from 2001:41d0:801:2000::539f: icmp_seq=1 ttl=46 time=20.9 ms

``` 

Drill also works : 

```  
drill aaaaaaaaaaaaaaaaaaaaaaaaa.aaaaaaaaaaaaaaaaaaaaaaaaa.taledo.games
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 15279
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 2, ADDITIONAL: 2 
;; QUESTION SECTION:
;; aaaaaaaaaaaaaaaaaaaaaaaaa.aaaaaaaaaaaaaaaaaaaaaaaaa.taledo.games.    IN      A

;; ANSWER SECTION:
aaaaaaaaaaaaaaaaaaaaaaaaa.aaaaaaaaaaaaaaaaaaaaaaaaa.taledo.games.       60      IN      A       51.195.216.156

;; AUTHORITY SECTION:
taledo.games.   3600    IN      NS      dns106.ovh.net.
taledo.games.   3600    IN      NS      ns106.ovh.net.

;; ADDITIONAL SECTION:
dns106.ovh.net. 70185   IN      A       213.251.188.150
ns106.ovh.net.  70185   IN      A       213.251.128.150

;; Query time: 429 msec
;; SERVER: 192.168.103.254
;; WHEN: Fri Nov  4 13:35:37 2022
;; MSG SIZE  rcvd: 178
``` 

Help is appreciated.

Cheers,
Tanguy

**Comment by tpelado** [comment]
Going further into this, it seems related to a truncated DNS request : 
![image](https://user-images.githubusercontent.com/25641016/199989739-705f7b14-7d40-4d8a-95f8-c6d002679ef5.png)

(see how it is missing the "s" on the TLD)

**Comment by bvanassche** [comment]
Thanks for the report. I think I know the root cause and I have started working on a fix.

**Comment by bvanassche** [comment]
Please retest with the latest version of the v5.9 branch or master branch. Although this issue has been marked as closed, further feedback is welcome.

**Comment by tpelado** [comment]
![Screenshot_20221108_102042](https://user-images.githubusercontent.com/25641016/200525660-ee43667e-9067-4744-9d01-9dbc82b57978.png)

Seems to be resolving fine with the latest version of the patch. Waiting now for package managers to pick up on this.

Thank you so much for your hard & swift job on this issue!

Cheers,

Tanguy

**Comment by basak** [comment]
Hi!

In Ubuntu we were asked to incorporate this fix in [Ubuntu bug 1998461](https://bugs.launchpad.net/ubuntu/+source/net-snmp/+bug/1998461), which led us to considering a cherry-pick of commit 49a0bca5a138a975ba503e6e8e0c7f1b72fe74be. In doing so, on review I had a couple of concerns. I wonder if you could take a look, please?

It looks like `netsnmp_parse_ep_str()` is called multiple times in succession, for example from `snmpIPv4BaseDomain.c` once for "default_target" and again against "inpeername" in the "Invalid default target" failure case. However:

a) While the patch adds various guards in the case that `ep_str.addr` is `NULL`, the error exit path in `netsnmp_parse_ep_str()` frees the pointer but does not reset `ep_str.addr` back to `NULL`. So what will happen later? Could this lead to an invalid memory access or double free?

b) Given that `netsnmp_parse_ep_str()` can be called multiple times, what happens when it allocates new memory to `ep_str.addr` again? Is this a memory leak?

It's not clear to me if either of the paths above are actually possible to trigger from outside the program. That would require further analysis to consider exactly these functions are called. But it seems to me that, at a minimum, the guards aren't complete unless all `free()` calls on the pointer reset the field to `NULL`, and that nothing new should be assigned to `ep_str.addr` before first freeing it if it was formerly not `NULL`. If these rules were followed then it would be much easier to verify the correctness of this fix.

Please could you take a look and see what you think? We'd appreciate your feedback before cherry-picking into Ubuntu. Thanks!

**Comment by sergiodj** [comment]
I noticed that this code is actually fixed in master: d420ff6cd3a3bd8ae469fc2d6a0bafd523280794 .

I think we must have backported the wrong patch (49a0bca5a138a975ba503e6e8e0c7f1b72fe74be actually doesn't belong to any branch, and is part of the initial PR).
